Convert checkbox and radio to React functional components

sfc-gh-tteixeira opened the pull request:
## Describe your changes

* Convert Checkbox and Radio widgets to React functional component
* Introduce `useBasicWidgetState()` to make it dead-easy to convert basic components to functional style
* Introduce `useFormClearHelper()` to make it dead-easy to use forms even in non-basic functional components

## GitHub Issue Link (if applicable)

n/a

## Testing Plan

- ~~Explanation of why no additional tests are needed~~
- Unit Tests (JS and/or Python)
- E2E Tests
- ~~Any manual testing needed?~~

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-tteixeira committed:
Checkbox checkpoint. Still broken.

sfc-gh-tteixeira committed:
Update Checkbox widget to functional component.

sfc-gh-tteixeira committed:
Introduce useBasicWidgetState and convert Checkbox and Radio to it.

sfc-gh-tteixeira committed:
Make useFormClearHelper follow latest pattern

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into checkbox

kmcgrady commented on frontend/lib/src/useBasicWidgetState.ts:
```diff
@@ -0,0 +1,129 @@
+
+    setCurrentValue(nextValueWSource.value)
+    updateWidgetMgrState(widgetMgr, nextValueWSource)
+  }, [nextValueWSource, setNextValueWSource, updateWidgetMgrState, widgetMgr])
```
I don't believe you need to add the setter into the setNextValueWSource, right?

kmcgrady commented on frontend/lib/src/components/widgets/Radio/Radio.tsx:
```diff
@@ -33,113 +33,73 @@ export interface Props {
+        element: RadioProto
+      ): RadioValue {
+        return widgetMgr.getIntValue(element)
+      },
```
Should these callbacks be wrapped in `useCallback`? Since they are used in the hook in the dependency arrays, these effects would be triggered every time, right?

sfc-gh-tteixeira replied:
Great catch. 

I just moved them to the module level instead of using useCallback. Feels cleaner.

kmcgrady commented on frontend/lib/src/components/widgets/Radio/Radio.tsx:
```diff
@@ -33,113 +33,73 @@ export interface Props {
 }
 
-export default Radio
+export default memo(Radio)
```
Making sure I understand this, memo ignores rerenders if the props haven't changed, which is a performance improvement. It seems the comparison is using [`Object.is`](https://react.dev/reference/react/memo#memo), which checks by reference for objects.  I'm curious how it would work for `element`. 

Since the element is coming from the delta message, I believe it's a new object, so this will always  rerender will continue the happen no matter what (and the memo, is probably not as useful).

So I guess the performance boost is from the other elements being modified and causing an (unnecessary rerender), is that correct? I imagine it might happen on page events (opening a dialog, or something).

sfc-gh-tteixeira replied:
That's right.

But also:
1. I think our default should be to use `memo` for every component, and only _not_ use it if there's a good reason.
2. At some point we should make the `element` object only update when it actually changed. This plus (1) above would save a ton of re-renders.

kmcgrady commented on frontend/lib/src/components/widgets/Checkbox/Checkbox.tsx:
```diff
@@ -56,250 +56,212 @@ interface ThemeProps {
 }
 
-export default withTheme(Checkbox)
+export default withTheme(memo(Checkbox))
```
Since we are a function component, we do not need the `withTheme` anymore, and we can call `useTheme` in the function instead.

sfc-gh-tteixeira replied:
Ohh, I wasn't aware of that. Thanks!

kmcgrady reviewed (commented):
Overall love the approach of the common hook for these widgets. Will love to combine these across all of our widgets.

Left a few comments with questions and suggestions. Let me know what you think!

sfc-gh-tteixeira committed:
Fix tests that required async/await

sfc-gh-tteixeira committed:
Address code review comments, lint, etc.

sfc-gh-tteixeira commented:
> Overall love the approach of the common hook for these widgets. Will love to combine these across all of our widgets.

I think we can refactor many of our widgets to use this new hook, but not all. Things like file uploader have their own peculiarities...

sfc-gh-tteixeira committed:
Forgot to rename props type

sfc-gh-tteixeira committed:
Make useFormClearHelper() use WidgetStateManager directly rather than FormClearHelper.

sfc-gh-tteixeira commented:
@kmcgrady PTAL again.

In addition to addressing your comments, I also:
* Made `useFormClearHelper()` use `WidgetStateManager` directly, as described [here](https://github.com/streamlit/streamlit/blob/develop/frontend/lib/src/components/widgets/Form/FormClearHelper.ts#L38).
* Converted the underlying `/shared/Radio` component to a functional component

sfc-gh-tteixeira committed:
Merge remote-tracking branch 'origin/develop' into checkbox

sfc-gh-tteixeira committed:
Also convert underlying /shared/Radio widget to a function component.

sfc-gh-tteixeira committed:
Fix radio test

kmcgrady commented on frontend/lib/src/components/shared/Radio/Radio.tsx:
```diff
@@ -43,172 +47,167 @@ export interface Props {
+  const theme = useTheme()
+  const { colors, radii } = theme
+  const style = { width }
+  const optionsCopy = [...options]
```
Curious what's the purpose of the copy? We are not modifying anything, right?

kmcgrady replied:
Ah I figured it out. Perhaps we should move this line to Line 96, so we can see that we make a copy because we _may_ push an element.

I'd probably re-organize to avoid modifying disabled and optionsCopy.

```javascript
const hasOptions = options.length > 0
// Either the user specified it as disabled or it's disabled because we don't have any options
const shouldDisable = disabled || !hasOptions
const optionsCopy = hasOptions ? options : ["No options to select."]

// ...
disabled={shouldDisable}
```

But that's probably not important for the refactor.

sfc-gh-tteixeira replied:
Yeah, there are plenty of places to improve stuff, but I didn't want to update too much.

That said, I just made the changes you proposed.

kmcgrady commented on frontend/lib/src/components/shared/Radio/Radio.tsx:
```diff
@@ -43,172 +47,167 @@ export interface Props {
-        return { value: prevProps.value ?? null }
-      })
+  useEffect(() => {
+    if (value === internalValue) {
```
nit: Might be worth a comment that `value` is the `defaultValue`, and when the default value changes that means it's a new widget?

Might be useful to actually rename the prop as `defaultValue` with:

`value: defaultValue`

for more clarity (and we don't need the word `internal`)

kmcgrady reviewed (approved):
Gave a couple suggestions and food for thought, but they do not impact the implementation details, so I'm approving it.

Looking forward to seeing `useBasicWidgetState` across all components. I also think we could benefit from sharing info like the `disabledColor`, etc, so we can avoid mismatching colors/other theming details across all our components.

sfc-gh-tteixeira committed:
Improve shared/Radio code nits.
